################################################################################
SHORT DESCRIPTION: 
################################################################################
Do not set event pending notifications for guests in S3.
"If domain is in S3 it will miss the notification so returns an error."

################################################################################
LONG DESCRIPTION: 
################################################################################
Appears to be in support of resuming PV Ubuntu guests from S3 when the host
is exiting S3.

################################################################################
CHANGELOG 
################################################################################
Patch was submitted to upstream Xen with the following sign-offs:
    Date: 10 May 2012
    From: Tomasz Wroblewski <tomasz.wroblewski@xxxxxxxxxx>
    Signed-off-by: Jean Guyader <jean.guyader@xxxxxxxxx>

################################################################################
REMOVAL 
################################################################################
see notes below.

################################################################################
UPSTREAM PLAN
################################################################################
This patch was submitted to upstream Xen on 10th of May 2012:
https://lists.xenproject.org/archives/html/xen-devel/2012-05/msg00585.html

It has not been merged and the thread did not continue.

OpenXT analysis in Sept 2016: https://openxt.atlassian.net/browse/OXT-578
noted the original ticket comment:
"""
Title: ubuntu PV s3 resume

Description: Works if is only performed in the guest; but not if performed in
conjunction with host s3, unsure why yet

Comment: I fixed it in xen by not setting the pending flag on evtchn if domain
is in S3 because the notification is missed.
This really needs some decent testing though"
"""

with OpenXT comment:
"""
We do not see why this is needed. Xen should know the guest is in S3
and not allow events to be sent. We are in favor of removing it.
"""

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/xen/common/event_2l.c
+++ b/xen/common/event_2l.c
@@ -18,6 +18,12 @@ static void evtchn_2l_set_pending(struct
     struct domain *d = v->domain;
     unsigned int port = evtchn->port;
 
+    /* if domain is in S3 it will miss the notification, so check here */
+    if (d->arch.hvm_domain.is_s3_suspended) {
+        return;
+    }
+
+
     /*
      * The following bit operations must happen in strict order.
      * NB. On x86, the atomic bit operations also act as memory barriers.
